### ISSUE TITLE
docker-machine / Openstack failing after host OS discovery (CentOS)

### ISSUE THREAD

**Issue by Didier-Lalli** [docker/machine#3089]
I’m hitting a problem with docker-machine/Openstack (version 0.6). I also tried going back to 0.5.5 and have the same behavior. I’m sure it’s something I’m doing wrong (client is CentOS7, cloud image is also CentOS), but I had it working fine back in November. I'm sitting behind corporate firewall, but I'm passing proxy details via --engine-env as shown below:
 
docker-machine --debug create --driver openstack  --openstack-insecure --openstack-ssh-user centos --openstack-image-name "CentOS7DockerReady"  --openstack-flavor-name m1.small --openstack-net-name "Demo Private Net1"  --openstack-floatingip-pool "external" --openstack-sec-groups DemoSecurityGroup --engine-env "HTTP_PROXY=http://proxy.corp.net:8088/"  --engine-env "HTTPS_PROXY=http://proxy.net:8088/"  docker-host1

Issue I have is that after the host OS is discovered (CentOS7), it fails with this: **open : no such file or directory**
But I have no information about what SSH command it’s attempting. Which is bizarre.
 
After it fails I can connect to the host with SSH, and continue manually the docker-engine installation.

Is there a more detailed log I could get to find out what went wrong?
Also could these “Couldn't set key” messages a problem?
 
Thanks a lot.
 
Didier

debug trace:
 
... 
Using SSH client type: external
{[-o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none centos@192.168.71.30 -o IdentitiesOnly=yes -i /root/.docker/machine/machines/docker-host1/id_rsa -p 22] /usr/bin/ssh}
About to run SSH command:
cat /etc/os-release
SSH cmd err, output: <nil>: NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
 
CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
 
 
Couldn't set key CPE_NAME, no corresponding struct field found
Couldn't set key , no corresponding struct field found
Couldn't set key CENTOS_MANTISBT_PROJECT, no corresponding struct field found
Couldn't set key CENTOS_MANTISBT_PROJECT_VERSION, no corresponding struct field found
Couldn't set key REDHAT_SUPPORT_PRODUCT, no corresponding struct field found
Couldn't set key REDHAT_SUPPORT_PRODUCT_VERSION, no corresponding struct field found
Couldn't set key , no corresponding struct field found
found compatible host: centos
Provisioning with centos...
(docker-host1) Calling .GetMachineName
(docker-host1) Calling .GetSSHHostname
(docker-host1) Calling .GetSSHPort
(docker-host1) Calling .GetSSHKeyPath
(docker-host1) Calling .GetSSHKeyPath
(docker-host1) Calling .GetSSHUsername
Using SSH client type: external
{[-o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none centos@192.168.71.30 -o IdentitiesOnly=yes -i /root/.docker/machine/machines/docker-host1/id_rsa -p 22] /usr/bin/ssh}
(docker-host1) Calling .GetSSHHostname
(docker-host1) Calling .GetSSHPort
(docker-host1) Calling .GetSSHKeyPath
(docker-host1) Calling .GetSSHKeyPath
(docker-host1) Calling .GetSSHUsername
Using SSH client type: external
{[-o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none centos@192.168.71.30 -o IdentitiesOnly=yes -i /root/.docker/machine/machines/docker-host1/id_rsa -p 22] /usr/bin/ssh}
installing base package: name=curl
(docker-host1) Calling .GetSSHHostname
(docker-host1) Calling .GetSSHPort
(docker-host1) Calling .GetSSHKeyPath
(docker-host1) Calling .GetSSHKeyPath
(docker-host1) Calling .GetSSHUsername
Using SSH client type: external
{[-o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none centos@192.168.71.30 -o IdentitiesOnly=yes -i /root/.docker/machine/machines/docker-host1/id_rsa -p 22] /usr/bin/ssh}
open : no such file or directory
notifying bugsnag: [Error running provisioning: exit status 1]
Error creating machine: Error running provisioning: exit status 1

**Comment by fsoppelsa** [comment]
Hi Didier, it looks like cloud-initi can't inject some values, and then probably something happens here https://github.com/docker/machine/blob/master/libmachine/provision/redhat_ssh_commander.go#L24

Is it 100% reproducible, and with the standard centos cloud image (qcow2, raw)?

@dgageot @nathanleclaire can you please add label OpenStack

**Comment by Didier-Lalli** [comment]
Hi Fabrizio, yes it's the standard CentOs cloud image (qcow2). And yes it's 100% reproducible, at least in my environment. Again it might be a problem on my side, so if anyone has this working, let me know.

**Comment by Didier-Lalli** [comment]
Ok, I made some progress. First of all I was able to get further in the install by adding the proxy settings in my profile and in my yum.conf (in addition to the --engine-env). 

I was then able to get to the docker install section but later on it failed after 10 attempts to find out if my docker daemon was up. After looking this up I discovered that it generated a /etc/systemd/system/docker.service with an incorrect ExecStart line:

ExecStart=/usr/bin/docker -d -H tcp://0.0.0.0:2376 -H unix:///var/run/docker.sock --storage-driver devicemapper --tlsverify --tlscacert /etc/docker/ca.pem --tlscert /etc/docker/server.pem --tls
key /etc/docker/server-key.pem --label provider=openstack

Feb 22 13:54:39 docker-host4 systemd[1]: Starting docker.service...
Feb 22 13:54:39 docker-host4 docker[2644]: flag provided but not defined: -d

-d doesn't seem to exist (anymore?) and daemon should be used instead.

Do I have a way to workaround this and force my own ExecStart?
